In [1]:
import numpy as np
import pandas as pd

The [O*NET Database](https://www.onetcenter.org/database.html) describes various jobs and the skills associated with them. Discovered from [a tweet by David Hood](https://twitter.com/Thoughtfulnz/status/1622029549310705664).

In [2]:
work_activities = pd.read_csv(
    "https://www.onetcenter.org/dl_files/database/db_27_1_text/Work%20Activities.txt",
    sep="\t",
)

In [3]:
work_activities

O*NET-SOC Code Element ID  \
0         11-1011.00  4.A.1.a.1   
1         11-1011.00  4.A.1.a.1   
2         11-1011.00  4.A.1.a.2   
3         11-1011.00  4.A.1.a.2   
4         11-1011.00  4.A.1.b.1   
...              ...        ...   
71581     53-7121.00  4.A.4.c.1   
71582     53-7121.00  4.A.4.c.2   
71583     53-7121.00  4.A.4.c.2   
71584     53-7121.00  4.A.4.c.3   
71585     53-7121.00  4.A.4.c.3   

                                           Element Name Scale ID  Data Value  \
0                                   Getting Information       IM        4.72   
1                                   Getting Information       LV        5.35   
2      Monitoring Processes, Materials, or Surroundings       IM        3.68   
3      Monitoring Processes, Materials, or Surroundings       LV        4.76   
4              Identifying Objects, Actions, and Events       IM        4.20   
...                                                 ...      ...         ...   
71581              Performing Administrative Activities       LV        2.27   
71582                     Staffing Organizational Units       IM        1.93   
71583                     Staffing Organizational Units       LV        1.60   
71584              Monitoring and Controlling Resources       IM        2.56   
71585              Monitoring and Controlling Resources       LV        2.64   

          N  Standard Error  Lower CI Bound  Upper CI Bound  \
0      35.0            0.13            4.46            4.98   
1      35.0            0.23            4.89            5.81   
2      35.0            0.17            3.33            4.03   
3      34.0            0.27            4.22            5.30   
4      35.0            0.15            3.91            4.50   
...     ...             ...             ...             ...   
71581  27.0            0.32            1.61            2.92   
71582  27.0            0.21            1.50            2.37   
71583  27.0            0.30            0.99            2.21   
71584  27.0            0.26            2.03            3.09   
71585  27.0            0.34            1.96            3.33   

      Recommend Suppress Not Relevant     Date Domain Source  
0                      N          NaN  07/2014     Incumbent  
1                      N            N  07/2014     Incumbent  
2                      N          NaN  07/2014     Incumbent  
3                      N            N  07/2014     Incumbent  
4                      N          NaN  07/2014     Incumbent  
...                  ...          ...      ...           ...  
71581                  N            N  08/2019     Incumbent  
71582                  N          NaN  08/2019     Incumbent  
71583                  N            N  08/2019     Incumbent  
71584                  N          NaN  08/2019     Incumbent  
71585                  N            N  08/2019     Incumbent  

[71586 rows x 13 columns]

The `O*NET-SOC Code` is the code for the occupation, the `Element Name` is the name of the work activity. The XLS version of the work activities file also contains the name of the occupation which is missing from the text file, but we don't need it right now.

Each row of the `work_activities` dataframe contains information on the *importance* (`IM`) or the *level* (`LV`) of a given work activity for a given occupation. The importance of the activity is related to how often that activity is needed for the occupation, whereas the level indicates the degree of skill needed when employing that activity.

For example, here's the first row of `work_activities`:

In [4]:
work_activities[["O*NET-SOC Code", "Element Name", "Scale ID", "Data Value"]].iloc[
    0:1, :
]

O*NET-SOC Code         Element Name Scale ID  Data Value
0     11-1011.00  Getting Information       IM        4.72

This says that for occupation `11-1011.00` (Chief Executives), the activity of "Getting Information" has an Importance of 4.72. The second row:

In [5]:
work_activities[["O*NET-SOC Code", "Element Name", "Scale ID", "Data Value"]].iloc[
    1:2, :
]

O*NET-SOC Code         Element Name Scale ID  Data Value
1     11-1011.00  Getting Information       LV        5.35

says that the level of the skill needed is 5.35. You need to also be aware that the importance has a range of 1-5, and level has a range of 0-7. For more on this see <https://www.onetonline.org/help/online/scales>. There are some other descriptors relevant to the importance and level including sample size, confidence intervals, and a recommendation of whether to not use the value, but I am just going to include all the data.

To get all the data for a given occupation into one row, we are going to need to `pivot` the data:

In [6]:
data = work_activities[
    ["O*NET-SOC Code", "Element Name", "Scale ID", "Data Value"]
].pivot(
    index="O*NET-SOC Code", columns=["Element Name", "Scale ID"], values="Data Value"
)
data

Element Name   Getting Information        \
Scale ID                        IM    LV   
O*NET-SOC Code                             
11-1011.00                    4.72  5.35   
11-1011.03                    4.78  5.37   
11-1021.00                    4.26  4.14   
11-2011.00                    4.32  4.67   
11-2021.00                    4.33  4.98   
...                            ...   ...   
53-7071.00                    4.38  3.87   
53-7072.00                    4.23  3.66   
53-7073.00                    3.54  3.45   
53-7081.00                    3.28  2.41   
53-7121.00                    4.40  4.00   

Element Name   Monitoring Processes, Materials, or Surroundings        \
Scale ID                                                     IM    LV   
O*NET-SOC Code                                                          
11-1011.00                                                 3.68  4.76   
11-1011.03                                                 3.48  4.07   
11-1021.00                                                 4.07  4.31   
11-2011.00                                                 2.78  3.08   
11-2021.00                                                 3.27  3.62   
...                                                         ...   ...   
53-7071.00                                                 4.23  4.96   
53-7072.00                                                 4.47  4.75   
53-7073.00                                                 4.11  4.50   
53-7081.00                                                 3.87  3.93   
53-7121.00                                                 4.62  5.37   

Element Name   Identifying Objects, Actions, and Events        \
Scale ID                                             IM    LV   
O*NET-SOC Code                                                  
11-1011.00                                         4.20  4.91   
11-1011.03                                         3.85  4.54   
11-1021.00                                         4.14  4.17   
11-2011.00                                         3.60  3.88   
11-2021.00                                         4.24  4.56   
...                                                 ...   ...   
53-7071.00                                         4.34  4.58   
53-7072.00                                         4.61  4.62   
53-7073.00                                         4.06  3.98   
53-7081.00                                         4.24  3.58   
53-7121.00                                         3.99  4.48   

Element Name   Inspecting Equipment, Structures, or Materials        \
Scale ID                                                   IM    LV   
O*NET-SOC Code                                                        
11-1011.00                                               2.38  2.11   
11-1011.03                                               2.11  2.37   
11-1021.00                                               3.43  3.14   
11-2011.00                                               1.77  1.08   
11-2021.00                                               1.45  0.72   
...                                                       ...   ...   
53-7071.00                                               4.18  4.30   
53-7072.00                                               4.90  4.69   
53-7073.00                                               4.34  4.21   
53-7081.00                                               4.60  4.01   
53-7121.00                                               4.56  4.74   

Element Name   Estimating the Quantifiable Characteristics of Products, Events, or Information  \
Scale ID                                                                                    IM   
O*NET-SOC Code                                                                                   
11-1011.00                                                   3.59                                
11-1011.03                                                   3

The result of that `pivot` is a multi-level index, so I am going to flatten it by merging the activity name (e.g. `Getting Information`) with either the `LV` or `IM` name to give e.g. `Getting Information LV`:

In [7]:
data.columns = [" ".join(col) for col in data.columns]
data

Getting Information IM  Getting Information LV  \
O*NET-SOC Code                                                   
11-1011.00                        4.72                    5.35   
11-1011.03                        4.78                    5.37   
11-1021.00                        4.26                    4.14   
11-2011.00                        4.32                    4.67   
11-2021.00                        4.33                    4.98   
...                                ...                     ...   
53-7071.00                        4.38                    3.87   
53-7072.00                        4.23                    3.66   
53-7073.00                        3.54                    3.45   
53-7081.00                        3.28                    2.41   
53-7121.00                        4.40                    4.00   

                Monitoring Processes, Materials, or Surroundings IM  \
O*NET-SOC Code                                                        
11-1011.00                                                   3.68     
11-1011.03                                                   3.48     
11-1021.00                                                   4.07     
11-2011.00                                                   2.78     
11-2021.00                                                   3.27     
...                                                           ...     
53-7071.00                                                   4.23     
53-7072.00                                                   4.47     
53-7073.00                                                   4.11     
53-7081.00                                                   3.87     
53-7121.00                                                   4.62     

                Monitoring Processes, Materials, or Surroundings LV  \
O*NET-SOC Code                                                        
11-1011.00                                                   4.76     
11-1011.03                                                   4.07     
11-1021.00                                                   4.31     
11-2011.00                                                   3.08     
11-2021.00                                                   3.62     
...                                                           ...     
53-7071.00                                                   4.96     
53-7072.00                                                   4.75     
53-7073.00                                                   4.50     
53-7081.00                                                   3.93     
53-7121.00                                                   5.37     

                Identifying Objects, Actions, and Events IM  \
O*NET-SOC Code                                                
11-1011.00                                             4.20   
11-1011.03                                             3.85   
11-1021.00                                             4.14   
11-2011.00                                             3.60   
11-2021.00                                             4.24   
...                                                     ...   
53-7071.00                                             4.34   
53-7072.00                                             4.61   
53-7073.00                                             4.06   
53-7081.00                                             4.24   
53-7121.00                                             3.99   

                Identifying Objects, Actions, and Events LV  \
O*NET-SOC Code                                                
11-1011.00                                             4.91   
11-1011.03                                             4.54   
11-1021.00                                             4.17   
11-2011.00                                             3.88   
11-2021.00                                             4.56   
...                                             

The occupation title represented by the `O*NET-SOC Code` can be found in a separate text file, which we will grab now:

In [8]:
occupation_data = pd.read_csv(
    "https://www.onetcenter.org/dl_files/database/db_27_1_text/Occupation%20Data.txt",
    sep="\t",
)
occupation_data = occupation_data.set_index("O*NET-SOC Code")
occupation_data

Title  \
O*NET-SOC Code                                                      
11-1011.00                                       Chief Executives   
11-1011.03                          Chief Sustainability Officers   
11-1021.00                        General and Operations Managers   
11-1031.00                                            Legislators   
11-2011.00                    Advertising and Promotions Managers   
...                                                           ...   
55-3014.00                     Artillery and Missile Crew Members   
55-3015.00                 Command and Control Center Specialists   
55-3016.00                                               Infantry   
55-3018.00                                         Special Forces   
55-3019.00      Military Enlisted Tactical Operations and Air/...   

                                                      Description  
O*NET-SOC Code                                                     
11-1011.00      Determine and formulate policies and provide o...  
11-1011.03      Communicate and coordinate with management, sh...  
11-1021.00      Plan, direct, or coordinate the operations of ...  
11-1031.00      Develop, introduce, or enact laws and statutes...  
11-2011.00      Plan, direct, or coordinate advertising polici...  
...                                                           ...  
55-3014.00      Target, fire, and maintain weapons used to des...  
55-3015.00      Operate and monitor communications, detection,...  
55-3016.00      Operate weapons and equipment in ground combat...  
55-3018.00      Implement unconventional operations by air, la...  
55-3019.00      All military enlisted tactical operations and ...  

[1016 rows x 2 columns]

There are also other taxonomic details present in the code, which represents a hierarchy. For exact details of what goes into the 2019 `O*NET-SOC Code` see <https://www.onetcenter.org/taxonomy/2019/structure.html>. Apart from wanting the specific occupation encoded by the full code, I will also extract the "major" group (represented by the first two numbers of the code before the dash), and the "minor" groups, which are indicated by the two digits after the dash (in combination with the major group). It represents a hierarchy of all the different job titles. For example, `11-1011.00`, the `Chief Executives`, are part of the `11` major group (Management Occupations) and the `10` minor group (Top Executives).

In [9]:
major_group_codes = {
    "11": "Management Occupations",
    "13": "Business and Financial Operations Occupations",
    "15": "Computer and Mathematical Occupations",
    "17": "Architecture and Engineering Occupations",
    "19": "Life, Physical, and Social Science Occupations",
    "21": "Community and Social Service Occupations",
    "23": "Legal Occupations",
    "25": "Educational Instruction and Library Occupations",
    "27": "Arts, Design, Entertainment, Sports, and Media Occupations",
    "29": "Healthcare Practitioners and Technical Occupations",
    "31": "Healthcare Support Occupations",
    "33": "Protective Service Occupations",
    "35": "Food Preparation and Serving Related Occupations",
    "37": "Building and Grounds Cleaning and Maintenance Occupations",
    "39": "Personal Care and Service Occupations",
    "41": "Sales and Related Occupations",
    "43": "Office and Administrative Support Occupations",
    "45": "Farming, Fishing, and Forestry Occupations",
    "47": "Construction and Extraction Occupations",
    "49": "Installation, Maintenance, and Repair Occupations",
    "51": "Production Occupations",
    "53": "Transportation and Material Moving Occupations",
    "55": "Military Specific Occupations",
}
minor_group_codes = {
    "11-10": "Top Executives",
    "11-20": "Advertising, Marketing, Promotions, Public Relations, and Sales Managers",
    "11-30": "Operations Specialties Managers",
    "11-90": "Other Management Occupations",
    "13-10": "Business Operations Specialists",
    "13-20": "Financial Specialists",
    "15-12": "Computer Occupations",
    "15-20": "Mathematical Science Occupations",
    "17-10": "Architects, Surveyors, and Cartographers",
    "17-20": "Engineers",
    "17-30": "Drafters, Engineering Technicians, and Mapping Technicians",
    "19-10": "Life Scientists",
    "19-20": "Physical Scientists",
    "19-30": "Social Scientists and Related Workers",
    "19-40": "Life, Physical, and Social Science Technicians",
    "19-50": "Occupational Health and Safety Specialists and Technicians",
    "21-10": "Counselors, Social Workers, and Other Community and Social Service Specialists",
    "21-20": "Religious Workers",
    "23-10": "Lawyers, Judges, and Related Workers",
    "23-20": "Legal Support Workers",
    "25-10": "Postsecondary Teachers",
    "25-20": "Preschool, Elementary, Middle, Secondary, and Special Education Teachers",
    "25-30": "Other Teachers and Instructors",
    "25-40": "Librarians, Curators, and Archivists",
    "25-90": "Other Educational Instruction and Library Occupations",
    "27-10": "Art and Design Workers",
    "27-20": "Entertainers and Performers, Sports and Related Workers",
    "27-30": "Media and Communication Workers",
    "27-40": "Media and Communication Equipment Workers",
    "29-10": "Healthcare Diagnosing or Treating Practitioners",
    "29-20": "Health Technologists and Technicians",
    "29-90": "Other Healthcare Practitioners and Technical Occupations",
    "31-11": "Home Health and Personal Care Aides; and Nursing Assistants, Orderlies, and Psychiatric Aides",
    "31-20": "Occupational Therapy and Physical Therapist Assistants and Aides",
    "31-90": "Other Healthcare Support Occupations",
    "33-10": "Supervisors of Protective Service Workers",
    "33-20": "Firefighting and Prevention Workers",
    "33-30": "Law Enforcement Workers",
    "33-90": "Other Protective Service Workers",
    "35-10": "Supervisors of Food Preparation and Serving Workers",
    "35-20": "Cooks and Food Preparation Workers",
    "35-30": "Food and Beverage Serving Workers",
    "35-90": "Other Food Preparation and Serving Related Workers",
    "37-10": "Supervisors of Building and Grounds Cleaning and Maintenance Workers",
    "37-20": "Building Cleaning and Pest Control Workers",
    "37-30": "Grounds Maintenance Workers",
    "39-10": "Supervisors of Personal Care and Service Workers",
    "39-20": "Animal Care and Service Workers",
    "39-30": "Entertainment Attendants and Related Workers",
    "39-40": "Funeral Service Workers",
    "39-50": "Personal Appearance Workers",
    "39-60": "Baggage Porters, Bellhops, and Concierges",
    "39-70": "Tour and Travel Guides",
    "39-90": "Other Personal Care and Service Workers",
    "41-10": "Supervisors of Sales Workers",
    "41-20": "Retail Sales Workers",
    "41-30": "Sales Representatives, Services",
    "41-40": "Sales Representatives, Wholesale and Manufacturing",
    "41-90": "Other Sales and Related Workers",
    "43-10": "Supervisors of Office and Administrative Support Workers",
    "43-20": "Communications Equipment Operators",
    "43-30": "Financial Clerks",
    "43-40": "Information and Record Clerks",
    "43-50": "Material Recording, Scheduling, Dispatching, and Distributing Workers",
    "43-60": "Secretaries and Administrative Assistants",
    "43-90": "Other Office and Administrative Support Workers",
    "45-10": "Supervisors of Farming, Fishing, and Forestry Workers",
    "45-20": "Agricultural Workers",
    "45-30": "Fishing and Hunting Workers",
    "45-40": "Forest, Conservation, and Logging Workers",
    "47-10": "Supervisors of Construction and Extraction Workers",
    "47-20": "Construction Trades Workers",
    "47-30": "Helpers, Construction Trades",
    "47-40": "Other Construction and Related Workers",
    "47-50": "Extraction Workers",
    "49-10": "Supervisors of Installation, Maintenance, and Repair Workers",
    "49-20": "Electrical and Electronic Equipment Mechanics, Installers, and Repairers",
    "49-30": "Vehicle and Mobile Equipment Mechanics, Installers, and Repairers",
    "49-90": "Other Installation, Maintenance, and Repair Occupations",
    "51-10": "Supervisors of Production Workers",
    "51-20": "Assemblers and Fabricators",
    "51-30": "Food Processing Workers",
    "51-40": "Metal Workers and Plastic Workers",
    "51-51": "Printing Workers",
    "51-60": "Textile, Apparel, and Furnishings Workers",
    "51-70": "Woodworkers",
    "51-80": "Plant and System Operators",
    "51-90": "Other Production Occupations",
    "53-10": "Supervisors of Transportation and Material Moving Workers",
    "53-20": "Air Transportation Workers",
    "53-30": "Motor Vehicle Operators",
    "53-40": "Rail Transportation Workers",
    "53-50": "Water Transportation Workers",
    "53-60": "Other Transportation Workers",
    "53-70": "Material Moving Workers",
    "55-10": "Military Officer Special and Tactical Operations Leaders",
    "55-20": "First-Line Enlisted Military Supervisors",
    "55-30": "Military Enlisted Tactical Operations and Air/Weapons Specialists and Crew Members",
}

Either I don't understand how the minor groups are calculated or not all occupations have a minor group, e.g. `11-3111.00` refers to `Compensation and Benefits Managers` but there isn't a `11-31` minor group. So we'll just set those to `None`:

In [10]:
occupation_data["Minor Group"] = occupation_data.index.to_series().apply(
    lambda x: minor_group_codes.get(x[:5], "None")
)

In [11]:
occupation_data["Major Group"] = occupation_data.index.to_series().apply(
    lambda x: major_group_codes[x[:2]]
)

In [12]:
data.shape, occupation_data.shape

((873, 82), (1016, 4))

Let's categorize the Minor Group and Major Group columns:

In [13]:
for col in ["Minor Group", "Major Group"]:
    occupation_data[col] = occupation_data[col].astype("category")

Before exporting, I will filter out the rows from the `occupation_data` which aren't in the `data`,
based on their shared O*NET-SOC code:

In [14]:
filtered_occupation_data = occupation_data.loc[data.index]
filtered_occupation_data.shape

(873, 4)

## Pipeline

In [15]:
from drnb.io.pipeline import create_default_pipeline

data_pipe = create_default_pipeline(
    check_for_duplicates=True,
    scale="z",
)

data_result = data_pipe.run(
    "onet271",
    data=data,
    target=filtered_occupation_data,
    url="https://www.onetcenter.org/database.html",
    tags=["small", "lowdim"],
    verbose=True,
)

[21:25:03] INFO     Requesting one extra neighbor to account for self-neighbor       ]8;id=844790;file:///home/james/dev/drnb3/src/drnb/neighbors/__init__.py\__init__.py]8;;\:]8;id=838191;file:///home/james/dev/drnb3/src/drnb/neighbors/__init__.py#623\623]8;;\

           INFO     Initial data shape: (873, 82)                                     ]8;id=194283;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=872051;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#83\83]8;;\

           INFO     Keeping all columns                                             ]8;id=586948;file:///home/james/dev/drnb3/src/drnb/preprocess.py\preprocess.py]8;;\:]8;id=387109;file:///home/james/dev/drnb3/src/drnb/preprocess.py#90\90]8;;\

           INFO     Data shape after filtering columns: (873, 82)                    ]8;id=757712;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=56937;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#164\164]8;;\

           INFO     Removing rows with NAs                                           ]8;id=243871;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=517551;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#145\145]8;;\

           INFO     Data shape after filtering NAs: (873, 82)                        ]8;id=748908;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=930513;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#156\156]8;;\

           INFO     Checked for duplicates: found 0                                  ]8;id=44798;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=638407;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#172\172]8;;\

           INFO     Z-Scaling                                                       ]8;id=969819;file:///home/james/dev/drnb3/src/drnb/preprocess.py\preprocess.py]8;;\:]8;id=613783;file:///home/james/dev/drnb3/src/drnb/preprocess.py#51\51]8;;\

           INFO     Converting to numpy with {'dtype': 'float32', 'layout': 'c'}     ]8;id=307920;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=725939;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#187\187]8;;\

           INFO     Writing data for onet271                                         ]8;id=21895;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=261245;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#298\298]8;;\

           INFO     Processing target with initial shape (873, 4)                    ]8;id=932966;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=635014;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#243\243]8;;\

           INFO     Keeping all columns                                             ]8;id=313191;file:///home/james/dev/drnb3/src/drnb/preprocess.py\preprocess.py]8;;\:]8;id=820942;file:///home/james/dev/drnb3/src/drnb/preprocess.py#90\90]8;;\

           INFO     Writing target for onet271                                       ]8;id=272425;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=223498;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#298\298]8;;\

           INFO     Calculating nearest neighbors                                    ]8;id=949151;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=331820;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#317\317]8;;\

           INFO     Loading faiss with AVX2 support.                                   ]8;id=721765;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/faiss-1.9.0-py3.12.egg/faiss/loader.py\loader.py]8;;\:]8;id=864504;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/faiss-1.9.0-py3.12.egg/faiss/loader.py#110\110]8;;\

           INFO     Successfully loaded faiss with AVX2 support.                       ]8;id=923360;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/faiss-1.9.0-py3.12.egg/faiss/loader.py\loader.py]8;;\:]8;id=487023;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/faiss-1.9.0-py3.12.egg/faiss/loader.py#112\112]8;;\

           INFO     Using 'faiss' to find nearest neighbors                           ]8;id=797526;file:///home/james/dev/drnb3/src/drnb/neighbors/__init__.py\__init__.py]8;;\:]8;id=766185;file:///home/james/dev/drnb3/src/drnb/neighbors/__init__.py#90\90]8;;\

           INFO     Finding 151 neighbors using faiss with euclidean metric and      ]8;id=948711;file:///home/james/dev/drnb3/src/drnb/neighbors/__init__.py\__init__.py]8;;\:]8;id=344572;file:///home/james/dev/drnb3/src/drnb/neighbors/__init__.py#103\103]8;;\
                    params: {}                                                                      

[21:25:04] INFO     Writing numpy format to                                          ]8;id=451269;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=321756;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#433\433]8;;\
                    nn/onet271.16.euclidean.exact.faiss.idx.npy                                     

           INFO     Writing csv format to                                            ]8;id=912495;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=588685;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#405\405]8;;\
                    nn/onet271.16.euclidean.exact.faiss.idx.csv                                     

           INFO     Writing numpy format to                                          ]8;id=166030;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=93661;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#433\433]8;;\
                    nn/onet271.16.euclidean.exact.faiss.dist.npy                                    

           INFO     Writing csv format to                                            ]8;id=66801;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=468370;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#405\405]8;;\
                    nn/onet271.16.euclidean.exact.faiss.dist.csv                                    

           INFO     Writing numpy format to                                          ]8;id=266088;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=794741;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#433\433]8;;\
                    nn/onet271.51.euclidean.exact.faiss.idx.npy                                     

           INFO     Writing csv format to                                            ]8;id=602485;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=925834;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#405\405]8;;\
                    nn/onet271.51.euclidean.exact.faiss.idx.csv                                     

           INFO     Writing numpy format to                                          ]8;id=558116;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=844752;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#433\433]8;;\
                    nn/onet271.51.euclidean.exact.faiss.dist.npy                                    

           INFO     Writing csv format to                                            ]8;id=904303;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=686970;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#405\405]8;;\
                    nn/onet271.51.euclidean.exact.faiss.dist.csv                                    

           INFO     Writing numpy format to                                          ]8;id=113650;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=950158;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#433\433]8;;\
                    nn/onet271.151.euclidean.exact.faiss.idx.npy                                    

           INFO     Writing csv format to                                            ]8;id=189638;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=853775;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#405\405]8;;\
                    nn/onet271.151.euclidean.exact.faiss.idx.csv                                    

           INFO     Writing numpy format to                                          ]8;id=788987;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=771486;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#433\433]8;;\
                    nn/onet271.151.euclidean.exact.faiss.dist.npy                                   

           INFO     Writing csv format to                                            ]8;id=257448;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=222102;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#405\405]8;;\
                    nn/onet271.151.euclidean.exact.faiss.dist.csv                                   

[21:25:05] INFO     Calculating triplets                                             ]8;id=803113;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=395651;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#331\331]8;;\

           INFO     Writing csv format to triplets/onet271.5.1337.idx.csv            ]8;id=360310;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=965754;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#405\405]8;;\

           INFO     Writing csv format to triplets/onet271.5.1337.euclidean.csv      ]8;id=61185;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=674019;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#405\405]8;;\

           INFO     Writing numpy format to triplets/onet271.5.1337.idx.npy          ]8;id=371094;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=594381;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#433\433]8;;\

           INFO     Writing numpy format to triplets/onet271.5.1337.euclidean.npy    ]8;id=229857;file:///home/james/dev/drnb3/src/drnb/io/__init__.py\__init__.py]8;;\:]8;id=433563;file:///home/james/dev/drnb3/src/drnb/io/__init__.py#433\433]8;;\

           INFO     Writing pipeline result for onet271                              ]8;id=198482;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py\pipeline.py]8;;\:]8;id=259880;file:///home/james/dev/drnb3/src/drnb/io/pipeline.py#134\134]8;;\

I decided to check whether this version of the data with the preparation in this notebook also placed jailers close to teachers with t-SNE. If you don't have the `jupyterlab-plotly` extension installed, in the `extra_plot` dict, set `clickable=False` and add `renderer="iframe"`.

In [1]:
import drnb.embed.pipeline as pl

pl.standard_eval(
    method="tsne",
    dataset="onet271",
    extra_plot=dict(
        plot="plotly",
        figsize=(10, 7),
        hover=["Minor Group", "Title"],
        alpha_scale=0.8,
        cex=30,
        clickable=True,
    ),
)

[23:12:56] INFO     KNN index provided. Ignoring KNN-related parameters.             ]8;id=182906;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/openTSNE/affinity.py\affinity.py]8;;\:]8;id=86708;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/openTSNE/affinity.py#190\190]8;;\

           INFO     Precomputed affinities provided. Ignoring perplexity-related        ]8;id=823355;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/openTSNE/tsne.py\tsne.py]8;;\:]8;id=871096;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/openTSNE/tsne.py#1340\1340]8;;\
                    parameters.                                                                     

           INFO     Precomputed initialization provided. Ignoring initalization-related ]8;id=113663;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/openTSNE/tsne.py\tsne.py]8;;\:]8;id=265250;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/openTSNE/tsne.py#1348\1348]8;;\
                    parameters.                                                                     

           INFO     Automatically determined negative gradient method `bh`                ]8;id=568130;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/openTSNE/tsne.py\tsne.py]8;;\:]8;id=721063;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/openTSNE/tsne.py#53\53]8;;\

[23:13:00] INFO     Automatically determined negative gradient method `bh`                ]8;id=543059;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/openTSNE/tsne.py\tsne.py]8;;\:]8;id=626298;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/openTSNE/tsne.py#53\53]8;;\

[23:13:09] INFO     Loading faiss with AVX2 support.                                   ]8;id=290938;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/faiss-1.9.0-py3.12.egg/faiss/loader.py\loader.py]8;;\:]8;id=715279;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/faiss-1.9.0-py3.12.egg/faiss/loader.py#110\110]8;;\

           INFO     Successfully loaded faiss with AVX2 support.                       ]8;id=811403;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/faiss-1.9.0-py3.12.egg/faiss/loader.py\loader.py]8;;\:]8;id=274887;file:///home/james/dev/drnb3/.venv/lib/python3.12/site-packages/faiss-1.9.0-py3.12.egg/faiss/loader.py#112\112]8;;\

FigureWidget({
    'data': [{'customdata': array([[122, 122, 'Architects, Surveyors, and Cartographers',
                                    'Architects, Except Landscape and Naval'],
                                   [123, 123, 'Architects, Surveyors, and Cartographers',
                                    'Landscape Architects'],
                                   [124, 124, 'Architects, Surveyors, and Cartographers',
                                    'Cartographers and Photogrammetrists'],
                                   ...,
                                   [174, 174,
                                    'Drafters, Engineering Technicians, and Mapping Technicians',
                                    'Non-Destructive Testing Specialists'],
                                   [175, 175,
                                    'Drafters, Engineering Technicians, and Mapping Technicians',
                                    'Photonics Technicians'],
                                 

[EvalResult(eval_type='RTE', label='rte-5-euclidean', value=np.float64(0.7718213058419243), info={'metric': 'euclidean', 'ntpp': 5}),
 EvalResult(eval_type='RPC', label='rpc-5-euclidean', value=np.float32(0.72743744), info={'metric': 'euclidean', 'ntpp': 5}),
 EvalResult(eval_type='NNP', label='nnp-15-noself-euclidean', value=np.float64(0.4938526155021), info={'metric': 'euclidean', 'n_neighbors': 15}),
 EvalResult(eval_type='NNP', label='nnp-50-noself-euclidean', value=np.float64(0.5157388316151202), info={'metric': 'euclidean', 'n_neighbors': 50}),
 EvalResult(eval_type='NNP', label='nnp-150-noself-euclidean', value=np.float64(0.5861779305078274), info={'metric': 'euclidean', 'n_neighbors': 150})]

The orange points are the educational occupations, while the beige points are the protective 
services occupations. In the latter group, jobs with the title "Correctional Officers and Jailers" 
were sadly *not* particularly close to any of the teacher jobs. They *are* close neighbors of 
Pharmacists, for reasons which are not apparent to me. Meanwhile some teacher jobs do end up near 
the Animal Trainers (although they aren't very close neighbors in the ambient space).

In [ ]:
from drnb.io.dataset import read_target

target = read_target("onet271")

In [3]:
target["Major Group"].cat.categories

Index(['Architecture and Engineering Occupations',
       'Arts, Design, Entertainment, Sports, and Media Occupations',
       'Building and Grounds Cleaning and Maintenance Occupations',
       'Business and Financial Operations Occupations',
       'Community and Social Service Occupations',
       'Computer and Mathematical Occupations',
       'Construction and Extraction Occupations',
       'Educational Instruction and Library Occupations',
       'Farming, Fishing, and Forestry Occupations',
       'Food Preparation and Serving Related Occupations',
       'Healthcare Practitioners and Technical Occupations',
       'Healthcare Support Occupations',
       'Installation, Maintenance, and Repair Occupations',
       'Legal Occupations', 'Life, Physical, and Social Science Occupations',
       'Management Occupations', 'Military Specific Occupations',
       'Office and Administrative Support Occupations',
       'Personal Care and Service Occupations', 'Production Occupations',
 